The time is too long to call `save_to_storage` if the saver is saving the latest checkpoint.

```text
[2024-01-17 19:16:54,785] [ERROR] [ckpt_saver.py:508:_save_shard] The step 15740 in event is no equal to step 15730 in memory.
[2024-01-17 19:16:54,785] [ERROR] [ckpt_saver.py:508:_save_shard] The step 15740 in event is no equal to step 15730 in memory.
[2024-01-17 19:16:54,785] [ERROR] [ckpt_saver.py:508:_save_shard] The step 15740 in event is no equal to step 15730 in memory.
[2024-01-17 19:16:54,785] [ERROR] [ckpt_saver.py:508:_save_shard] The step 15740 in event is no equal to step 15730 in memory.
[2024-01-17 19:16:54,785] [ERROR] [ckpt_saver.py:814:save_step_checkpoint] Rank 0 save checkpoint failed for step 15740
[2024-01-17 19:16:54,785] [ERROR] [ckpt_saver.py:508:_save_shard] The step 15740 in event is no equal to step 15730 in memory.
[2024-01-17 19:16:54,786] [ERROR] [ckpt_saver.py:508:_save_shard] The step 15740 in event is no equal to step 15730 in memory.
[2024-01-17 19:16:54,786] [ERROR] [ckpt_saver.py:508:_save_shard] The step 15740 in event is no equal to step 15730 in memory.
[2024-01-17 19:16:54,786] [ERROR] [ckpt_saver.py:508:_save_shard] The step 15740 in event is no equal to step 15730 in memory.
[2024-01-17 19:16:54,786] [ERROR] [ckpt_saver.py:814:save_step_checkpoint] Rank 1 save checkpoint failed for step 15740
[2024-01-17 19:16:54,786] [ERROR] [ckpt_saver.py:814:save_step_checkpoint] Rank 2 save checkpoint failed for step 15740
[2024-01-17 19:16:54,786] [ERROR] [ckpt_saver.py:814:save_step_checkpoint] Rank 3 save checkpoint failed for step 15740
[2024-01-17 19:16:54,786] [ERROR] [ckpt_saver.py:814:save_step_checkpoint] Rank 4 save checkpoint failed for step 15740
[2024-01-17 19:16:54,786] [ERROR] [ckpt_saver.py:814:save_step_checkpoint] Rank 5 save checkpoint failed for step 15740
[2024-01-17 19:16:54,786] [ERROR] [ckpt_saver.py:814:save_step_checkpoint] Rank 6 save checkpoint failed for step 15740
[2024-01-17 19:16:54,786] [ERROR] [ckpt_saver.py:814:save_step_checkpoint] Rank 7 save checkpoint failed for step 15740
[2024-01-17 19:16:54,786] [ERROR] [ckpt_saver.py:822:save_step_checkpoint] Rank 0 save checkpoint failed for step 15740
```

--- a/dlrover/python/common/multi_process.py
+++ b/dlrover/python/common/multi_process.py
@@ -38,9 +38,11 @@ def retry_socket(func):
     def wrapper(self, *args, **kwargs):
         retry = kwargs.get("retry", 30)
         succeed = False
-        for _ in range(retry):
+        for i in range(retry):
             try:
-                return func(self, *args, **kwargs)
+                result = func(self, *args, **kwargs)
+                succeed = True
+                return result
             except (FileNotFoundError, ConnectionRefusedError):
                 time.sleep(1)
         if not succeed:

--- a/dlrover/python/common/multi_process.py
+++ b/dlrover/python/common/multi_process.py
@@ -34,6 +34,21 @@ SUCCESS_CODE = "OK"
 ERROR_CODE = "ERROR"
 
 
+def retry_socket(func):
+    def wrapper(self, *args, **kwargs):
+        retry = kwargs.get("retry", 30)
+        succeed = False
+        for _ in range(retry):
+            try:
+                return func(self, *args, **kwargs)
+            except (FileNotFoundError, ConnectionRefusedError):
+                time.sleep(1)
+        if not succeed:
+            return func(self, *args, **kwargs)
+
+    return wrapper
+
+
 def _create_socket_server(path):
     """
     Create a socket server.
@@ -197,6 +212,7 @@ class LocalSocketComm(metaclass=ABCMeta):
         """Synchronize the obj between processes."""
         pass
 
+    @retry_socket
     def _request(self, request: SocketRequest):
         """Create a socket client to requet the shared object."""
         client = _create_socket_client(self._socket_file)

--- a/dlrover/python/elastic_agent/torch/ckpt_saver.py
+++ b/dlrover/python/elastic_agent/torch/ckpt_saver.py
@@ -918,11 +918,9 @@ class TempDirCheckpointSaver(AsyncCheckpointSaver):
 
                 # commit checkpoint
                 shutil.move(tmp_path, target_path)
-
-                self.update_tracker_file(step)
-
                 # clean stage dir
                 shutil.rmtree(step_done_dir)
+                self.update_tracker_file(step)
                 logger.info(
                     f"Commit checkpoint tmp_path: {tmp_path}, "
                     f"path: {target_path}"

--- a/dlrover/python/master/scaler/pod_scaler.py
+++ b/dlrover/python/master/scaler/pod_scaler.py
@@ -410,8 +410,11 @@ class PodScaler(Scaler):
 
         # The two env vars is compatible with kubeflow/PytorchJob because
         # users may use the scripts of kubeflow/PytorchJob in the ElasticJob.
-        env.append(V1EnvVar(name=NodeEnv.WORLD_SIZE, value=str(worker_num)))
-        env.append(V1EnvVar(name=NodeEnv.RANK, value=str(node.rank_index)))
+        if self._distribution_strategy == DistributionStrategy.ALLREDUCE:
+            env.append(
+                V1EnvVar(name=NodeEnv.WORLD_SIZE, value=str(worker_num))
+            )
+            env.append(V1EnvVar(name=NodeEnv.RANK, value=str(node.rank_index)))
 
         # Deprecated env vars
         env.append(V1EnvVar(name=NodeEnv.WORKER_TYPE, value=node.type))

--- a/dlrover/python/tests/test_multi_process.py
+++ b/dlrover/python/tests/test_multi_process.py
@@ -19,10 +19,22 @@ from dlrover.python.common.multi_process import (
     SharedLock,
     SharedMemory,
     SharedQueue,
+    retry_socket,
 )
 
 
-class SharedLockTest(unittest.TestCase):
+class SocketTest(object):
+    @retry_socket
+    def test_retry(self, retry):
+        raise FileNotFoundError("test")
+
+
+class SharedObjectTest(unittest.TestCase):
+    def test_retry(self):
+        t = SocketTest()
+        with self.assertRaises(FileNotFoundError):
+            t.test_retry(retry=1)
+
     def test_shared_lock(self):
         name = "test"
         os.environ["TORCHELASTIC_RUN_ID"] = "test_job"

--- a/dlrover/python/tests/test_pod_scaler.py
+++ b/dlrover/python/tests/test_pod_scaler.py
@@ -124,6 +124,20 @@ class PodScalerTest(unittest.TestCase):
         self.assertEqual(pod.spec.volumes[0].name, "pvc-nas")
         self.assertEqual(len(main_container.volume_mounts), 1)
 
+        scaler._distribution_strategy = DistributionStrategy.ALLREDUCE
+        node = Node(NodeType.WORKER, 0, resource, rank_index=0)
+        pod = scaler._create_pod(node, pod_stats, [])
+        main_container = pod.spec.containers[0]
+        world_size = -1
+        rank = -1
+        for env in main_container.env:
+            if env.name == "WORLD_SIZE":
+                world_size = int(env.value)
+            elif env.name == "RANK":
+                rank = int(env.value)
+        self.assertEqual(world_size, 2)
+        self.assertEqual(rank, 0)
+
     def test_create_service(self):
         scaler = PodScaler("elasticjob-sample", "default")
         scaler.start()

--- a/dlrover/trainer/tests/torch/deepspeed_ckpt_test.py
+++ b/dlrover/trainer/tests/torch/deepspeed_ckpt_test.py
@@ -129,10 +129,13 @@ class DeepSpeedCheckpointTest(unittest.TestCase):
                 tmpdirname, str(step), storage_type=StorageType.DISK
             )
             # Wait asynchronously saving.
+            start = time.time()
             while True:
                 if "100" in os.listdir(tmpdirname):
                     break
                 time.sleep(0.1)
+                if time.time() - start > 10:
+                    break
 
             with open(tracer_file, "r") as f:
                 restored_step = int(f.read())

--- a/dlrover/trainer/torch/flash_checkpoint/ddp_engine.py
+++ b/dlrover/trainer/torch/flash_checkpoint/ddp_engine.py
@@ -127,9 +127,9 @@ class DdpCheckpointEngine(CheckpointEngine):
         succeed = False
         if step > self._cached_step:
             succeed = self.save_to_memory(step, state_dict, paths)
+        # Only rank 0 persist the checkpoint to the storage.
         if succeed and self._rank == 0:
             event = CheckpointEvent(type=CheckpointEventType.SAVE, step=step)
-            # Only rank 0 persist the checkpoint to the storage.
             self._event_queue.put(event)
 
     def load(self, resume_path=""):

--- a/dlrover/trainer/torch/flash_checkpoint/ddp_engine.py
+++ b/dlrover/trainer/torch/flash_checkpoint/ddp_engine.py
@@ -124,7 +124,7 @@ class DdpCheckpointEngine(CheckpointEngine):
         """
         if self._local_rank != 0:
             return
-        succeed = False
+        succeed = True
         if step > self._cached_step:
             succeed = self.save_to_memory(step, state_dict, paths)
         # Only rank 0 persist the checkpoint to the storage.

--- a/dlrover/trainer/torch/flash_checkpoint/ddp_engine.py
+++ b/dlrover/trainer/torch/flash_checkpoint/ddp_engine.py
@@ -104,7 +104,7 @@ class DdpCheckpointEngine(CheckpointEngine):
                 the value is the path of storage to save.
         """
         conf = CheckpointConfig(step=step, paths=paths)
-        self.save_state_dict_to_memory(state_dict, conf)
+        return self.save_state_dict_to_memory(state_dict, conf)
 
     @timer
     def save_to_storage(self, step, state_dict, paths):
@@ -124,12 +124,13 @@ class DdpCheckpointEngine(CheckpointEngine):
         """
         if self._local_rank != 0:
             return
+        succeed = False
         if step > self._cached_step:
-            self.save_to_memory(step, state_dict, paths)
-        event = CheckpointEvent(type=CheckpointEventType.SAVE, step=step)
-
-        # Only rank 0 persist the checkpoint to the storage.
-        self._event_queue.put(event)
+            succeed = self.save_to_memory(step, state_dict, paths)
+        if not succeed and self._rank == 0:
+            event = CheckpointEvent(type=CheckpointEventType.SAVE, step=step)
+            # Only rank 0 persist the checkpoint to the storage.
+            self._event_queue.put(event)
 
     def load(self, resume_path=""):
         """

--- a/dlrover/trainer/torch/flash_checkpoint/ddp_engine.py
+++ b/dlrover/trainer/torch/flash_checkpoint/ddp_engine.py
@@ -127,7 +127,7 @@ class DdpCheckpointEngine(CheckpointEngine):
         succeed = False
         if step > self._cached_step:
             succeed = self.save_to_memory(step, state_dict, paths)
-        if not succeed and self._rank == 0:
+        if succeed and self._rank == 0:
             event = CheckpointEvent(type=CheckpointEventType.SAVE, step=step)
             # Only rank 0 persist the checkpoint to the storage.
             self._event_queue.put(event)

--- a/dlrover/trainer/torch/flash_checkpoint/deepspeed_engine.py
+++ b/dlrover/trainer/torch/flash_checkpoint/deepspeed_engine.py
@@ -90,7 +90,7 @@ class DeepSpeedCheckpointEngine(CheckpointEngine):
                 the value is the path of storage to save.
         """
         conf = CheckpointConfig(step=step, paths=paths)
-        self.save_state_dict_to_memory(state_dict, conf)
+        return self.save_state_dict_to_memory(state_dict, conf)
 
     @timer
     def save_to_storage(self, step, state_dict, paths):
@@ -108,13 +108,14 @@ class DeepSpeedCheckpointEngine(CheckpointEngine):
                 ["model_states", "optim_states"] of the state dict and
                 the value is the path of storage to save.
         """
+        succeed = False
         if step > self._cached_step:
-            self.save_to_memory(step, state_dict, paths)
+            succeed = self.save_to_memory(step, state_dict, paths)
 
         # Only local rank 0 to notify the saving event to the agent.
         if self._local_rank != 0:
             return
-        if state_dict:
+        if succeed:
             event = CheckpointEvent(type=CheckpointEventType.SAVE, step=step)
             self._event_queue.put(event)
 

--- a/dlrover/trainer/torch/flash_checkpoint/deepspeed_engine.py
+++ b/dlrover/trainer/torch/flash_checkpoint/deepspeed_engine.py
@@ -113,11 +113,10 @@ class DeepSpeedCheckpointEngine(CheckpointEngine):
             succeed = self.save_to_memory(step, state_dict, paths)
 
         # Only local rank 0 to notify the saving event to the agent.
-        if self._local_rank != 0:
+        if self._local_rank != 0 and not succeed:
             return
-        if succeed:
-            event = CheckpointEvent(type=CheckpointEventType.SAVE, step=step)
-            self._event_queue.put(event)
+        event = CheckpointEvent(type=CheckpointEventType.SAVE, step=step)
+        self._event_queue.put(event)
 
     def get_local_shard_num(self):
         local_world_size = env_utils.get_local_world_size()

--- a/dlrover/trainer/torch/flash_checkpoint/deepspeed_engine.py
+++ b/dlrover/trainer/torch/flash_checkpoint/deepspeed_engine.py
@@ -108,15 +108,14 @@ class DeepSpeedCheckpointEngine(CheckpointEngine):
                 ["model_states", "optim_states"] of the state dict and
                 the value is the path of storage to save.
         """
-        succeed = False
+        succeed = True
         if step > self._cached_step:
             succeed = self.save_to_memory(step, state_dict, paths)
 
         # Only local rank 0 to notify the saving event to the agent.
-        if self._local_rank != 0 and not succeed:
-            return
-        event = CheckpointEvent(type=CheckpointEventType.SAVE, step=step)
-        self._event_queue.put(event)
+        if self._local_rank == 0 and succeed:
+            event = CheckpointEvent(type=CheckpointEventType.SAVE, step=step)
+            self._event_queue.put(event)
 
     def get_local_shard_num(self):
         local_world_size = env_utils.get_local_world_size()

--- a/dlrover/trainer/torch/flash_checkpoint/engine.py
+++ b/dlrover/trainer/torch/flash_checkpoint/engine.py
@@ -14,6 +14,7 @@
 import os
 import time
 from abc import ABCMeta, abstractmethod
+from datetime import timedelta
 from multiprocessing import Process
 from typing import Dict
 
@@ -42,8 +43,11 @@ def check_all_rank_ready(group: dist.ProcessGroup, ready):
     """
     if not group:
         return ready
+    backend = dist.get_backend()
+    local_rank = env_utils.get_local_rank()
+    device = "cpu" if backend == "gloo" else f"cuda:{local_rank}"
     value = 0 if ready else 1
-    t = torch.tensor([value], dtype=torch.int64)
+    t = torch.tensor([value], dtype=torch.int32).to(device)
     dist.all_reduce(t, group=group)
     return t == 0
 
@@ -54,9 +58,12 @@ def verify_all_rank_step_consistent(group: dist.ProcessGroup, step):
     """
     if not group:
         return True
-    t = torch.Tensor([float(step)])
+    backend = dist.get_backend()
+    local_rank = env_utils.get_local_rank()
+    device = "cpu" if backend == "gloo" else f"cuda:{local_rank}"
+    t = torch.tensor([float(step)]).to(device)
     world_size = group.size()
-    outputs = [torch.Tensor([0.0]) for _ in range(world_size)]
+    outputs = [torch.tensor([0.0]) for _ in range(world_size)]
     dist.all_gather(outputs, t, group=group)
     for step in outputs:
         if not torch.equal(step, outputs[0]):
@@ -129,7 +136,11 @@ class CheckpointEngine(metaclass=ABCMeta):
         self.storage = storage
         if dist.is_initialized():
             self._rank = dist.get_rank()
-            self._loader_group = dist.new_group(backend="gloo")
+            backend = dist.get_backend()
+            self._loader_group = dist.new_group(
+                backend=backend,
+                timeout=timedelta(seconds=30),
+            )
         else:
             self._rank = 0
             self._loader_group = None
@@ -225,7 +236,7 @@ class CheckpointEngine(metaclass=ABCMeta):
 
     def save_state_dict_to_memory(self, state_dict, conf: CheckpointConfig):
         if self._local_rank != self.local_shard_id:
-            return
+            return False
 
         acquired = self._shm_lock.acquire(blocking=False)
         all_rank_ready = check_all_rank_ready(self._saver_group, acquired)
@@ -237,7 +248,7 @@ class CheckpointEngine(metaclass=ABCMeta):
             )
             if acquired:
                 self._shm_lock.release()
-            return
+            return False
         state_dict[DLROVER_CKPT_CONFIG_KEY] = conf
         self._shm_handler.save_state_dict(state_dict)
 
@@ -246,6 +257,7 @@ class CheckpointEngine(metaclass=ABCMeta):
         self._cached_step = conf.step
         if dist.is_initialized():
             dist.barrier(group=self._saver_group)
+        return True
 
     def get_state_dict_from_memory(self):
         state_dict = {}

--- a/dlrover/trainer/torch/flash_checkpoint/engine.py
+++ b/dlrover/trainer/torch/flash_checkpoint/engine.py
@@ -202,16 +202,7 @@ class CheckpointEngine(metaclass=ABCMeta):
             },
         )
 
-        succeed = False
-        for _ in range(3):
-            try:
-                queue.put(class_meta)
-                succeed = True
-                break
-            except FileNotFoundError:
-                time.sleep(3)
-        if not succeed:
-            queue.put(class_meta)
+        queue.put(class_meta)
         queue.unlink()
 
     def _update_saver_config(self):
@@ -226,12 +217,6 @@ class CheckpointEngine(metaclass=ABCMeta):
                 raise ValueError(
                     "The event queue cannot be None on local rank 0."
                 )
-            for _ in range(3):
-                try:
-                    self._event_queue.put(event)
-                    return
-                except FileNotFoundError:
-                    time.sleep(3)
             self._event_queue.put(event)
 
     def save_state_dict_to_memory(self, state_dict, conf: CheckpointConfig):

--- a/dlrover/trainer/torch/flash_checkpoint/fsdp_engine.py
+++ b/dlrover/trainer/torch/flash_checkpoint/fsdp_engine.py
@@ -441,7 +441,7 @@ class FsdpCheckpointEngine(CheckpointEngine):
                 only if the training process fails.
         """
         if self._local_rank != self.local_shard_id:
-            return
+            return False
 
         acquired = self._shm_lock.acquire(blocking=False)
         all_rank_ready = check_all_rank_ready(self._saver_group, acquired)
@@ -453,7 +453,7 @@ class FsdpCheckpointEngine(CheckpointEngine):
             )
             if acquired:
                 self._shm_lock.release()
-            return
+            return False
 
         conf = CheckpointConfig(step=step)
         conf.writing_shm = True
@@ -481,6 +481,7 @@ class FsdpCheckpointEngine(CheckpointEngine):
         self._cached_step = conf.step
         if dist.is_initialized():
             dist.barrier(group=self._saver_group)
+        return True
 
     def save_to_storage(self, step, state_dict, paths: Dict[str, str]):
         """
@@ -490,18 +491,19 @@ class FsdpCheckpointEngine(CheckpointEngine):
             step (int): the iteration step.
             state_dict (dict): the state dict of model and optimizer to save.
         """
+        succeed = False
         if step > self._cached_step:
-            self.save_to_memory(step, state_dict, paths)
+            succeed = self.save_to_memory(step, state_dict, paths)
 
         # Only local rank 0 on each node notifies the event to save.
-        if self._local_rank != 0:
+        if self._local_rank != 0 and not succeed:
             return
-        if paths:
-            logger.info(
-                "Put a save event to notify the agent persists checkpoint."
-            )
-            event = CheckpointEvent(type=CheckpointEventType.SAVE, step=step)
-            self._event_queue.put(event)
+
+        logger.info(
+            "Put a save event to notify the agent persists checkpoint."
+        )
+        event = CheckpointEvent(type=CheckpointEventType.SAVE, step=step)
+        self._event_queue.put(event)
 
     def get_saver_class(self):
         """

--- a/dlrover/trainer/torch/flash_checkpoint/fsdp_engine.py
+++ b/dlrover/trainer/torch/flash_checkpoint/fsdp_engine.py
@@ -491,19 +491,17 @@ class FsdpCheckpointEngine(CheckpointEngine):
             step (int): the iteration step.
             state_dict (dict): the state dict of model and optimizer to save.
         """
-        succeed = False
+        succeed = True
         if step > self._cached_step:
             succeed = self.save_to_memory(step, state_dict, paths)
 
         # Only local rank 0 on each node notifies the event to save.
-        if self._local_rank != 0 and not succeed:
-            return
-
-        logger.info(
-            "Put a save event to notify the agent persists checkpoint."
-        )
-        event = CheckpointEvent(type=CheckpointEventType.SAVE, step=step)
-        self._event_queue.put(event)
+        if self._local_rank == 0 and succeed:
+            logger.info(
+                "Put a save event to notify the agent persists checkpoint."
+            )
+            event = CheckpointEvent(type=CheckpointEventType.SAVE, step=step)
+            self._event_queue.put(event)
 
     def get_saver_class(self):
         """

--- a/dlrover/trainer/torch/flash_checkpoint/megatron_engine.py
+++ b/dlrover/trainer/torch/flash_checkpoint/megatron_engine.py
@@ -99,7 +99,7 @@ class MegatronCheckpointEngine(CheckpointEngine):
             state_dict (dict): the state dict of model and optimizer to save.
         """
         conf = CheckpointConfig(step=step, paths=paths)
-        self.save_state_dict_to_memory(state_dict, conf)
+        return self.save_state_dict_to_memory(state_dict, conf)
 
     @timer
     def save_to_storage(self, step, state_dict, paths):
@@ -114,13 +114,14 @@ class MegatronCheckpointEngine(CheckpointEngine):
             step (int): the iteration step.
             state_dict (dict): the state dict of model and optimizer to save.
         """
+        succeed = False
         if step > self._cached_step:
-            self.save_to_memory(step, state_dict, paths)
+            succeed = self.save_to_memory(step, state_dict, paths)
 
         # Only local rank 0 to notify the saving event to the agent.
         if self._dp_rank != 0 or self._local_rank != 0:
             return
-        if state_dict:
+        if succeed:
             event = CheckpointEvent(type=CheckpointEventType.SAVE, step=step)
             self._event_queue.put(event)
 

--- a/dlrover/trainer/torch/flash_checkpoint/megatron_engine.py
+++ b/dlrover/trainer/torch/flash_checkpoint/megatron_engine.py
@@ -114,7 +114,7 @@ class MegatronCheckpointEngine(CheckpointEngine):
             step (int): the iteration step.
             state_dict (dict): the state dict of model and optimizer to save.
         """
-        succeed = False
+        succeed = True
         if step > self._cached_step:
             succeed = self.save_to_memory(step, state_dict, paths)
 

